src/Docfx.App/PdfBuilder.cs

```diff
@@ -94,27 +94,56 @@ public static async Task CreatePdf(string outputFolder)
 
         using var pageLimiter = new SemaphoreSlim(Environment.ProcessorCount, Environment.ProcessorCount);
         var pagePool = new ConcurrentBag<IPage>();
+
         var headerFooterCache = new ConcurrentDictionary<(string, string), Task<byte[]>>();
 
-        await AnsiConsole.Progress().StartAsync(async progress =>
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        using var ctr = SubscribeCancelKeyPressEvent(cts);
+
+        var pdfBuildTask = AnsiConsole.Progress().StartAsync(async progress =>
         {
-            await Parallel.ForEachAsync(pdfTocs, async (item, _) =>
+            await Parallel.ForEachAsync(pdfTocs, new ParallelOptions { CancellationToken = cts.Token }, async (item, _) =>
             {
                 var (url, toc) = item;
                 var outputName = Path.Combine(Path.GetDirectoryName(url) ?? "", toc.pdfFileName ?? Path.ChangeExtension(Path.GetFileName(url), ".pdf"));
                 var task = progress.AddTask(outputName);
-                var outputPath = Path.Combine(outputFolder, outputName);
+                var pdfOutputPath = Path.Combine(outputFolder, outputName);
 
                 await CreatePdf(
-                    PrintPdf, PrintHeaderFooter, task, new(baseUrl, url), toc, outputFolder, outputPath,
-                    pageNumbers => pdfPageNumbers[url] = pageNumbers);
+                    PrintPdf, PrintHeaderFooter, task, new(baseUrl, url), toc, outputFolder, pdfOutputPath,
+                    pageNumbers => pdfPageNumbers[url] = pageNumbers,
+                    cts.Token);
 
                 task.Value = task.MaxValue;
                 task.StopTask();
             });
         });
 
+        // Wait pdfBuildTask completed or cancelled.
+        await Task.WhenAny(pdfBuildTask, Task.Delay(Timeout.Infinite, cts.Token));
```

Comment:
Is there a reason we are not relying on the cancellation token passed to `pdfBuildTask` to gracefully cancel the PDF build job? This extra manual shutdown code looks suspicious to me.

Reply:
It's required because `AnsiConsole.Progress().StartAsync` and Playwright APIs doesn't support `CancellationToken` parameter.

So when operation is cancelled via CancellationToken.
It takes some seconds(about 5-10 seconds) until operations are cancelled.

So I need to use `Task.WhenAny`  for stopping operations immediately.
And additional dipose/await code are required.
Because It can't output logs until `AnsiConsole.Progress().StartAsync` is completed  
(Otherwise console log is outputted to wrong places)